### PR TITLE
Add constraints for our AES-GCM implementation

### PR DESCRIPTION
**Pull request by olorin**
Probably best to leave this until after whiteboard review tomorrow.

### CODE REVIEW

**Comment by olorin** [comment]
This is ready for review (ref https://github.com/ambiata/ops/pull/533#issuecomment-267237680).

! @erikd-ambiata @thumphries @novemberkilo

File: src/Tinfoil/AEAD/AESGCM/Constraints.hs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+maxCleartextLength =
+  Blocks 4294967294
+
+-- |
+-- Upper bound on number of messages encrypted with the same key. This is
+-- due to the length of the counter part of the invocation field of the IV,
+-- @2^32@ bits.
+maxMessageCount :: Messages
+maxMessageCount =
+  Messages 4294967296
```
Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: Same.

**Comment by erikd-ambiata** [comment]
:+1: with fix of magic numbers.

**Comment by thumphries** [comment]
🍯

File: src/Tinfoil/AEAD/AESGCM/Constraints.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+-- @2^32@ blocks; block size is @2^7@ bits, so we have @2^39@ bits worth of
+-- counter. We lose one block due to the GCM counter skipping 0, and another
+-- to encrypt the @GHASH@ with the first block in the keystream.
+--
+-- So in blocks, we have a limit of @(2^32 - 2)@.
+--
+-- <dx.doi.org/10.6028/NIST.SP.800-38D NIST SP 800-38D>
+maxCleartextLength :: Blocks
+maxCleartextLength =
+  Blocks 4294967294
```
Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: Thats just constant folding. I would be *extremely* surprised is GHC didn't do that one with `-O1` and above.

Comment (jystic) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: super easy to find out `mafia build --dump-core` then look in `dist/build/src/Tinfoil/AEAD/AESGCM/Constraints.dump-simpl`

Comment (olorin) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: Yeah, I thought I observed it not happening a while ago, but checking now it looks to be constant. I think the difference is last time I had a polymorphic function in the mix; I probably could have fixed it with a `SPECIALIZE` pragma.

Comment (jystic) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: (just make sure the file is touched so ghc actually has to build it)

Comment (jystic) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: fwiw, I usually don't trust it either :shrugs:

Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: Holy crap, you are right. It doesn't constant fold!

Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: I was using `-keep-s-file` which is perfect.

Comment (olorin) [inline src/Tinfoil/AEAD/AESGCM/Constraints.hs]: Ah handy, will remember that one. Thanks.

**Comment by olorin** [comment]
[hipchat] Decided to leave the magic numbers, as GHC does not actually constant-fold shifts. I may try and fix this some weekend, but this doesn't help now. Added comments clarifying reasoning.

**Pull request merged**